control-plane/pkg/prober/cache.go

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2021 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"container/list"
+	"context"
+	"sync"
+	"time"
+)
+
+const (
+	// StatusReady signals that a given object is ready.
+	StatusReady Status = iota
+	// StatusUnknown signals that a given object is not ready and its state is unknown.
+	StatusUnknown
+	// StatusNotReady signals that a given object is not ready.
+	StatusNotReady
+)
+
+// Status represents the resource status.
+type Status int
+
+// Cache is a key-status store.
+type Cache interface {
+	// GetStatus retries the status associated with the given key.
+	GetStatus(key string) Status
+	// UpsertStatus add or updates the status associated with the given key.
+	// Once the given key expires the onExpired callback will be called passing the arg parameter.
+	UpsertStatus(key string, status Status, arg interface{}, onExpired ExpiredFunc)
+}
+
+// ExpiredFunc is a callback called once an entry in the cache is expired.
+type ExpiredFunc func(key string, arg interface{})
+
+type inMemoryLocalCache struct {
+	// mu protects targets and entries
+	mu sync.RWMutex
+	// targets is a map of key-pointer to an element in the entries list.
+	targets map[string]*list.Element
+	// entries is a doubly-linked list of all entries present in the cache.
+	// This allows fast deletion of expired entries.
+	entries *list.List
+
+	expireDuration time.Duration
```

Comment:
expireDuration -> expiration ?

Reply:
isn't that typically called `ttl` ?

Reply:
not typically. I like `expiration` over `ttl`. I think expiring is _stronger_ than `ttl`

(heavily biased based on heavy usage of a sweet library like https://github.com/jhalterman/expiringmap - which this PR is bascially also trying to do)

Reply:
I'm using expiration, for now, I think we all get the meaning of this variable with that name.